app/lua/lua.c

```diff
@@ -22,498 +19,284 @@
+    /* likewise if not CR terminated, then unread and ditto */
+    lua_pushlightfunction(L, &pipe_unread);              /* pobj:read at S[1] */
+    lua_insert(L, 1);
+    lua_call(L, 2, 0);                                   /* pobj:unread(line) */
+    lua_pushboolean(L, false);
+    return 1;                                                 /* return false */
+  } else {
+//dbg_printf("popping CR terminated string(%d) %s", l-1,  b);
   }
-  return 0;
-}
+ /*
+  * Now we can process a proper CR terminated line
+  */
+  lua_pushlstring(L, b, --l);                                /* remove end CR */
+  lua_remove(L, 2);
+  b = lua_tostring(L, 2);
+
+  if (MLref != LUA_NOREF) {
+    /* processing multiline */
+    lua_rawgeti(L, LUA_REGISTRYINDEX, MLref);         /* insert prev lines(s) */
+    lua_pushliteral(L, "\n");                                    /* insert CR */
+    lua_pushvalue(L, 2);                                      /* dup new line */
+    lua_concat(L, 3);                                         /* concat all 3 */
+    lua_remove(L, 2);                               /* and shift down to S[2] */
+   } else if (b[0] == '=') {
+    /* If firstline and of the format =<expression> */
+    lua_pushfstring(L, "return %s", b+1);
+    lua_remove(L, 2);
+  }
+
+  /* ToS is at S[2] which contains the putative chunk to be compiled */
 
+  status = luaL_loadbuffer(L, lua_tostring(L, 2), lua_strlen(L, 2), "=stdin");
 
-static int runargs (lua_State *L, char **argv, int n) {
-  int i;
-  for (i = 1; i < n; i++) {
-    if (argv[i] == NULL) continue;
-    lua_assert(argv[i][0] == '-');
-    switch (argv[i][1]) {  /* option */
-      case 'e': {
-        const char *chunk = argv[i] + 2;
-        if (*chunk == '\0') chunk = argv[++i];
-        lua_assert(chunk != NULL);
-        if (dostring(L, chunk, "=(command line)") != 0)
-          return 1;
-        break;
-      }
-      case 'm': {
-        const char *limit = argv[i] + 2;
-        int memlimit=0;
-        if (*limit == '\0') limit = argv[++i];
-        lua_assert(limit != NULL);
-        memlimit = c_atoi(limit);
-        lua_gc(L, LUA_GCSETMEMLIMIT, memlimit);
-        break;
-      }
-      case 'l': {
-        const char *filename = argv[i] + 2;
-        if (*filename == '\0') filename = argv[++i];
-        lua_assert(filename != NULL);
-        if (dolibrary(L, filename))
-          return 1;  /* stop if file fails */
-        break;
-      }
-      default: break;
+  if (incomplete(L, status)) {
+    /* Store line back in the Reg mlref sot */
+    if (MLref == LUA_NOREF) {
+      MLref = luaL_ref(L, LUA_REGISTRYINDEX);
+    } else {
+      lua_rawseti(L, LUA_REGISTRYINDEX, MLref);
     }
+  } else {
+    /* compile finished OK or with hard error */
+    lua_remove(L, 2);                    /* remove line because now redundant */
+    if (MLref!= LUA_NOREF) {       /* also remove multiline if it exists */
+      luaL_unref(L, LUA_REGISTRYINDEX, MLref);
```

Comment:
I wish `luaL_unref()` returned `LUA_NOREF` so the unrefs could be one-liners easily.

Reply:
So do I but this sort of rationalisation is for a downstream PR.